AsyncDecodingSource.cpp

```diff
@@ -0,0 +1,408 @@
+        bool isVorbis)
+    : mComponentName(codecName),
+      mSource(source),
+      mLooper(looper),
+      mCodecLooper(new ALooper),
+      mMeta(new MetaData),
+      mUsingSurface(false),
+      mIsVorbis(isVorbis) {
+    mCodecLooper->setName("codec_looper");
+    mCodecLooper->start(false, false, ANDROID_PRIORITY_AUDIO);
+
+    mCodec = MediaCodec::CreateByComponentName(mCodecLooper, mComponentName);
+
+    mReflector = new AHandlerReflector<AsyncDecodingSource>(this);
+    mLooper->registerHandler(mReflector);
+    mNotify = new AMessage(0, mReflector);
+}
+
+AsyncDecodingSource::~AsyncDecodingSource() {
+    mCodec->release();
+    mCodecLooper->stop();
+    mLooper->unregisterHandler(mReflector->id());
+}
+
+bool AsyncDecodingSource::configure(const sp<AMessage> format, const sp<Surface> surface, uint32_t flags) {
+    mCodec->setCallback(mNotify);
+    status_t err = mCodec->configure(format, surface, NULL /* crypto */, flags);
+    if (err != OK) {
+        ALOGE("Failed to configure codec '%s'", mComponentName.c_str());
+        return false;
+    }
+    mUsingSurface = surface != NULL;
+
+    // Set output format metadata
+    sp<AMessage> outputFormat;
+    mCodec->getOutputFormat(&outputFormat);
+    sp<MetaData> meta = new MetaData;
+    convertMessageToMetaData(outputFormat, meta);
+    mMeta.lock().set(meta);
+
+    mOutput.lock()->mReachedEOS = false;
+    ALOGI("Configured codec '%s'!", mComponentName.c_str());
+    return true;
+}
+
+status_t AsyncDecodingSource::start(MetaData *params) {
+    (void)params;
+    Mutexed<Output>::Locked me(mOutput);
+
+    if (mState != INIT) {
+        return -EINVAL;
+    }
+
+    status_t res = mCodec->start();
+    if (res == OK) {
+        res = mSource->start();
+    }
+
+    if (res == OK) {
+        mState = STARTED;
+        me->mReachedEOS = false;
+    } else {
+        mState = ERROR;
+    }
+    return res;
+}
+
+status_t AsyncDecodingSource::stop() {
+    Mutexed<Output>::Locked me(mOutput);
+    if (mState != STARTED) {
+        return -EINVAL;
+    }
+
+    // wait for any pending reads to complete
+    mState = STOPPING;
+    while (me->mReading) {
+        me.waitForCondition(me->mReadCondition);
+    }
+
+    status_t res1 = mCodec->stop();
+    if (res1 != OK) {
+        mCodec->release();
+    }
+    status_t res2 = mSource->stop();
+    if (res1 == OK && res2 == OK) {
+        mState = STOPPED;
+    } else {
+        mState = ERROR;
+    }
+    return res1 != OK ? res1 : res2;
+}
+
+sp<MetaData> AsyncDecodingSource::getFormat() {
+    Mutexed<sp<MetaData>>::Locked meta(mMeta);
+    return *meta;
+}
+
+AsyncDecodingSource::Output::Output()
+    : mReachedEOS(false),
+      mReading(false) {
+}
+
+status_t AsyncDecodingSource::read(
+#if ANDROID_MAJOR >= 9
+                    MediaBufferBase **buffer,
+#else
+                    MediaBuffer **buffer,
+#endif
+                    const ReadOptions *options) {
+    *buffer = NULL;
```

Comment:
Any reason using NULL over nullptr?

Reply:
Just that it came from the original. Switched them all to nullptr now.